<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>

Юрий, доброго дня! рад приветствовать тебя)<br />
Сейчас большие выходные и большие праздники. С которыми тебя хотел бы тоже тебя поздравить!<br>
А также, у нас сейчас новогодние каникулы - но у меня было свободное время и я решил проверить твою работу. Не могу обещать повторную проверку в течении 24 часов до конца каникул. Но 11 января все ревьюеры начнут работать в  штатном режиме и я обязательно увижу твой проект снова.<br>
<br />
Меня зовут Николай Шавлюго. <br />И на этом этапе твоего движения к новой профессии от меня будут комментарии по написанному коду. <br />Чтобы меньше "мусорить эфир" и чтобы максимально наглядно отделяться от основного кода, есть предложение условиться в некоторых обозначениях:<br /> <br />
<u><b>ТАКОЙ ШРИФТ</b></u> - всегда начало комментария <br />
<font color='green'>такой шрифт</font> - комментарии о том, что всё ОК <br/>
<font color='orange'>такой шрифт</font> - комментарии о том, что всё ОК по результату,<br> однако есть на что обратить внимание в плане применения техник, или есть способы сделать более короткий или быстрый код <br/>
<font color='red'>такой шрифт</font> - комментарии о том, что есть критичный момент, влияющий на бизнес-результат проекта.<br/>
</font><br /><br />
<font color='green'>Моей целью является не "уличить" в не знании, а просто высказать сверху твоих знаний - свой опыт, что бы ты мог использовать его для своего дальнейшего успеха) И очень здорово будет, если тебе удастся задавать вопросы, да и вообще - всячески доставать меня, если я по каким-то причинам не приму проект:) При этом, из своего опыта скажу, хорошо и важно, когда переписка ревьюера и студента - сохраняется на следующие проекты и даже на будущую практическую деятельность.<br>
<br> Я сразу же начал общаться на ты:) но если это не очень удобно, то дай знать))<BR> В ПУТЬ!<br>

# Описание проекта

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

# 1. Загрузка и подготовка данных

In [1]:
#Импортируем данные
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
sns.set()



import warnings
warnings.filterwarnings('ignore')

from scipy import stats as st 
from sklearn import metrics
from sklearn.metrics import roc_curve, roc_auc_score, r2_score, mean_absolute_error
from sklearn.metrics import mean_squared_error # as mse
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.model_selection import train_test_split,cross_val_score, GridSearchCV, StratifiedShuffleSplit
import locale
locale.setlocale(locale.LC_ALL, '')
#g0 = pd.read_csv('/datasets/geo_data_0.csv')
#g1= pd.read_csv('/datasets/geo_data_1.csv')
#g2= pd.read_csv('/datasets/geo_data_2.csv')



'en_US.UTF-8'

In [2]:
g0 = pd.read_csv('/datasets/geo_data_0.csv')
g1= pd.read_csv('/datasets/geo_data_1.csv')
g2= pd.read_csv('/datasets/geo_data_2.csv')

### Сбросим столбец ID

In [3]:
g0 = g0.drop('id', axis=1)
g1 = g1.drop('id', axis=1)
g2 = g2.drop('id', axis=1)


### Проверим основную информацию по данным

In [4]:
g0.info()
g0.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB


f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647

In [5]:
g1.info()
g1.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB


f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305

In [6]:
g2.info()
g2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB


f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746

## Вывод 

С данными всё в порядке. Все имеют тип данных float64. Можем переходить к обучению и проверки моделей.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Данные на базе) ОК, хорошо, когда есть аналитическая культура - обязательно делать первичный анализ данных, не смотря на проект)

# 2. Обучение и проверка модели

Я буду поочередно обучать и проверять модели, по территориям.

## Первая модель

In [50]:
target0 = g0['product']
features0= g0.drop('product', axis=1)
numeric = ['f0', 'f1', 'f2']

In [51]:
state = np.random.RandomState(12345)

In [52]:
features_train0, features_valid0, target_train0, target_valid0 = train_test_split(
   features0, target0, test_size=0.25, random_state=state)
#scaler = StandardScaler()
#scaler.fit(features_train0[numeric])
#features_train0[numeric] = scaler.transform(features_train0[numeric])
#features_valid0[numeric] = scaler.transform(features_valid0[numeric])
model = LinearRegression()
model.fit(features_train0, target_train0)
predict0 = model.predict(features_valid0)
mse0 = mean_squared_error(target_valid0, predict0)
print("g0:")
print("RMSE =", mse0 ** 0.5)
print("mean =", predict0.mean())
print("R2 =", r2_score(target_valid0, predict0))


g0:
RMSE = 37.5794217150813
mean = 92.59256778438038
R2 = 0.27994321524487786


## Вторая модель

In [53]:
target1 = g1['product']
features1= g1.drop('product', axis=1)

In [54]:

features_train1, features_valid1, target_train1, target_valid1 = train_test_split(
   features1, target1, test_size=0.25, random_state=12345)
numeric = ['f0', 'f1', 'f2']
scaler.fit(features_train1[numeric])
features_train1[numeric] = scaler.transform(features_train1[numeric])
features_valid1[numeric] = scaler.transform(features_valid1[numeric])
model.fit(features_train1, target_train1)
predict1 = model.predict(features_valid1)
mse1 = mean_squared_error(target_valid1, predict1)
print("")
print("g1:")
print("RMSE =", mse1 ** 0.5)
print("mean =", predict1.mean())
print("R2 =", r2_score(target_valid1, predict1))


g1:
RMSE = 0.8930992867756158
mean = 68.728546895446
R2 = 0.9996233978805127


## Третья модель

In [55]:
target2 = g2['product']
features2= g2.drop('product', axis=1)

In [56]:
features_train2, features_valid2, target_train2, target_valid2 = train_test_split(
   features2, target2, test_size=0.25, random_state=12345)
scaler.fit(features_train2[numeric])
features_train2[numeric] = scaler.transform(features_train2[numeric])
features_valid2[numeric] = scaler.transform(features_valid2[numeric])
model.fit(features_train2, target_train2)
predict2 = model.predict(features_valid2)
mse2 = mean_squared_error(target_valid2, predict2)
print("")
print("g2:")
print("RMSE =", mse2 ** 0.5)
print("mean =", predict2.mean())
print("R2 =", r2_score(target_valid2, predict2))


g2:
RMSE = 40.02970873393434
mean = 94.96504596800489
R2 = 0.20524758386040443


## Вывод

В модели номер 2, метрика RMSE меньше всех. В 3 модели средний запас больше всего, однако RMSE хуже всего. Лучше всех показала себя g1, т.к был небольшой объем данных.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Отлично, с расчётами согласен.<br>
Единственное - все три региона имеют одинаковые расчёты. Следовательно можно автоматизировать и тем самым уменьшить количество кода)

# 3. Подготовка к расчёту прибыли

Исходя из полученных данных, можно сделать вывод, что любой регион подходит нам, и компания будет в плюсе.

In [57]:
POINT_PER_REGION = 500
BUDGET = 10000000000
COST_PER_WELL = 50000000
COST_ONE_BAREL = 450000
RISK = 0.025

In [58]:
BUDGET/COST_PER_WELL

200.0

In [59]:
def roi(revenue, investment):
    return (revenue-investment)/investment
    
roi_coef = 1    
for amount in range(1000, 1000000000000, 1):
    if roi(COST_ONE_BAREL*amount, BUDGET)>0:
        print('Необходимый запас барелей, чтобы произошла окупаемость 200 скважин =',amount)
        print('Необходимый запас барелей на одну скважину =', round(amount/200,0))
        print('ROI = {:.2f}'.format(roi(COST_ONE_BAREL*amount, BUDGET)))
        break

Необходимый запас барелей, чтобы произошла окупаемость 200 скважин = 22223
Необходимый запас барелей на одну скважину = 111.0
ROI = 0.00


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='red'><br> у нас должно быть 111 на скважину. При задании константы COST_ONE_BAREL надо сделать 450000
<font color='blue'><br> Сделано

In [60]:
# БЫЛО
pred_LR_reg1_s = pd.Series(predict0, index=range(25000))
pred_LR_reg2_s = pd.Series(predict1, index=range(25000))
pred_LR_reg3_s = pd.Series(predict2, index=range(25000))

y_test_1_s = pd.Series(target0).reset_index(drop=True)
y_test_2_s = pd.Series(target1).reset_index(drop=True)
y_test_3_s = pd.Series(target2).reset_index(drop=True)

In [66]:
# СТАЛО
pred_LR_reg1_s = pd.Series(predict0, index=range(25000))
pred_LR_reg2_s = pd.Series(predict1, index=range(25000))
pred_LR_reg3_s = pd.Series(predict2, index=range(25000))

y_test_1_s = pd.Series(target_valid0).reset_index(drop=True)
y_test_2_s = pd.Series(target_valid1).reset_index(drop=True)
y_test_3_s = pd.Series(target_valid2).reset_index(drop=True)

In [67]:
sample_pred = {\
               'predictions_linear_r1':pred_LR_reg1_s,\
               'predictions_linear_r2':pred_LR_reg2_s,\
               'predictions_linear_r3':pred_LR_reg3_s}
sample_y = {\
            'y_test_1':y_test_1_s,\
            'y_test_2':y_test_2_s,\
            'y_test_3':y_test_3_s}

zips = zip(sample_y.values(), sample_pred.values())
zips

### Необходимые функции для расчета

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='red'><br> 1) если выше скорректировать COST_ONE_BAREL, то нужно ли ниже умножать на 1000?<br>
2) может лучше здесь сразу оформить функцию прибыли? <br>
3) в функцию revenue в mean_revenue_and_std надо посылать target_subsample, а не target
<font color='blue'><br> Сделал

In [68]:
def revenue(target, probabilities, count):
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:count]
    return COST_ONE_BAREL * selected.sum()

# РЕВЬЮЕР: нужна ли здесь 1000? И второе: здесь не учитывается затраты, т.е. это не прибыль, а просто выручка.

def gross_profit(revenue, cost):
    return revenue - cost

def mean_revenue_and_std(target, probabilities, max_point_well=500, well=200, quantile=0.05):

    state = np.random.RandomState(12345)

    #храним выручку
    values = []
    for i in range(1000):
        #поиск средней выручки
        target_subsample = target.sample(n=max_point_well, replace=True, random_state=state)
        probs_subsample = probabilities[target_subsample.index]
        values.append(revenue(target_subsample, probs_subsample, well))

        # РЕВЬЮЕР: в функцию revenue надо посылать target_subsample, а не target
    
    values = pd.Series(values)
    
   # loss = len(values[values<0])/len(values)
   # print('loss',loss)
   
    loss = ((values-BUDGET)<0).mean()
    print('Risk',loss*100,'%')    
    
    
    lower = values.quantile(0.05)
    
    #Выручка
    mean = values.mean()
    print("Средняя выручка:", locale.format('%d', mean, grouping=True))
    print("Средняя выручка 2:", mean)
    
    #Коэффициент возврата инвестиций
    roi_coef = abs(roi(BUDGET, mean))
    print('roi_coef', roi_coef)
    
    #нижняя граница доверительного интервала прибыли - 95%
    lower_private = values.quantile(.025) - BUDGET 
    print('lower_private',locale.format('%d', lower_private, grouping=True))
    
    #верхняя граница доверительного интервала прибыли - 95%
    upper_private = values.quantile(.975) - BUDGET   
    print('upper_private',locale.format('%d', upper_private, grouping=True))
    
    gross_prof = gross_profit(mean, BUDGET) 
    print('gross_prof',locale.format('%d', gross_prof, grouping=True))

    return values, mean, lower_private, upper_private, gross_prof, roi_coef

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">

<br><font color='BLUE'> 
<b><u>КОММЕНТАРИЙ УЧЕНИКА</u></b>
</font><br>
<font color='red'><br>
Переделал блок с расчетами, поэтому удалил предыдущий код. Новый вариант, я считаю, более структурированным и правильным.

# 4. Расчёт прибыли и рисков 

Расчитывать прибыль буду так-же, деля на регионы.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 

<br><font color='BLUE'> 
<b><u>КОММЕНТАРИЙ УЧЕНИКА</u></b>
</font><br>
<font color='red'><br>
Опять таки, тут тоже удалил предыдущие шаги и полностью переделал анализ.

In [69]:

reg = 0
revenue_df = pd.DataFrame(columns=['net_conf_down','gross_profit', 'net_conf_up', 'revenue', 'ROI'])

for y, prob_y in zips:
    reg+=1
    print('\nРегион '+str(reg)+':')
    values, mean, d, up , gross, r = mean_revenue_and_std(y, prob_y);
    revenue_df.loc[reg] = [d, gross, up, mean, r]
        
pd.options.display.float_format = '{:,.2f}'.format    
revenue_df


Регион 1:
Risk 6.0 %
Средняя выручка: 10,425,938,526
Средняя выручка 2: 10425938526.910593
roi_coef 0.04085373473200468
lower_private -102,090,094
upper_private 947,976,353
gross_prof 425,938,526

Регион 2:
Risk 1.0 %
Средняя выручка: 10,515,222,773
Средняя выручка 2: 10515222773.44329
roi_coef 0.04899779914739524
lower_private 68,873,225
upper_private 931,547,591
gross_prof 515,222,773

Регион 3:
Risk 6.4 %
Средняя выручка: 10,435,008,362
Средняя выручка 2: 10435008362.782757
roi_coef 0.04168739953618503
lower_private -128,880,547
upper_private 969,706,954
gross_prof 435,008,362


net_conf_down   gross_profit    net_conf_up           revenue  ROI
1 -102,090,094.84 425,938,526.91 947,976,353.36 10,425,938,526.91 0.04
2   68,873,225.37 515,222,773.44 931,547,591.26 10,515,222,773.44 0.05
3 -128,880,547.33 435,008,362.78 969,706,954.18 10,435,008,362.78 0.04

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font><br>
<font color='red'><br> прибыль получается на уровне 50 -70 млрд.. Это круто, но многовато...<br>
Для ориентира - прибыль по региону 2 - около 450-520 млн. остальные - чуть меньше. Т.е. не более 500 млн., если в целом.<br>
Некоторые моменты увидел и оставил в тексте комментарии - посмотри, пожалуйста.
<font color='blue'><br> Сделал твои изменения. Теперь gross_profit отрицательный). пересмотри еще раз.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 3</u></b>
</font><br>
<font color='red'><br>
Привет, Юрий!  Посмотри, пожалуйста, выше код:<br>
1) Там было описка в ячейке 60<br>
2) Плюс расчёт рисков - чуть скорректировал.
    
Буду ждать тогда твоих комментариев и выводов по проекту (кто победитель?:)

In [ ]:
************************************************

In [ ]:
*******************************************

# Финальный вывод

Исходя из наших расчетов и анализируя показатель чистой прибыли с каждого региона, можно сделать следующие выводы. Регион 1 и 3 приблизительно равны по предполагаемой чистой прибыли. Выделяется регион номер 2 - его чистая прибыль равна 512 млн. Следуя нашему условию, мы можем выбрать этот регион и признать его выигрышным.

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 4</u></b>
</font><br>
<font color='green'><br>ОК, Юрий, именно так)<br>
Спасибо за проект) УСПЕШНОГО ДАЛЬНЕЙШЕГО ОБУЧЕНИЯ!

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font><br>
<font color='green'><br>
Юрий, на самом деле очень последовательная работа!<br>
И аргументация, и удобный для просмотра код, и графика - многое удалось как надо<br>
      
Данный проект не простой, может показаться. Потому что, его цель не просто построить прогноз по модели или прогноз продаж. А научиться выбирать лучшее управленческое решение на основе метода моделирования ситуации. В этом ключе и есть смысл её рассматривать. (Кстати, одна из программ на рынке по этой теме - AnyLogic - интересный продукт. Мне приходилось как-то смотерть её.. интересно. Можешь ролик из инета глянуть по какому-нибудь примеру с её использованием. Тогда тема управленческого моделирования может по-другому лечь... Ведь бутсреп (в лоджике егое нет) - как раз один из спобов моделирования)<br>    

<font color='red'>    
Есть несколько моментов, которые надо бы проверить:<br>
1) Расчёт функций прибыли и убытков: в тексте оставил алгоритм - посмотри, пожалуйта.<br>
2) также, нужен расчёт рисков: это % количества случаев расчёта, имеющих отрицательное значение прибыли. И расчёт доверительных интервалов.

<font color='green'>    
Спасибо)<br>
Буду ждать комментариев)
<br>
</font>
<font color='blue'>    
Спасибо тебе за помощь<br>
Максимально постарался переделать проект и уделил ему больше времени. Думаю, в этот раз уже всё в порядке :). Впереди жесткий дедлайн 13 января и у меня как раз остается два дня подумать над сборным проектом и сдать его. Надеюсь, еще удастся поработать с тобой вместе над следующими проектами, спасибо тебе большое еще раз за помощь.
<br>
</font>    

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
<font color='green'> 
    <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font><br>
<font color='red'><br>>Юрий, привет!<br>
Грандиозную работу ещё провёл. Мне очень бы хотелось принято работу..<br>
Но не могу, пока...<br>
Посмотри, пожалуйста, те комментарии, что я оставил в тексте. Возможно, что благодаря им удастся получить ожидаемую прибыль. (не более 500 млн.!)<br>

Постараюсь сразу брать твою работу в провреку, помня о дедлайне.

# Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован